src/snowflake/snowpark/_internal/temp_table_cleaner.py

```diff
@@ -0,0 +1,78 @@
+#
+# Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
+#
+
+import logging
+import threading
+from collections import defaultdict
+from queue import Queue
+from typing import TYPE_CHECKING, Dict
+
+if TYPE_CHECKING:
+    from snowflake.snowpark.session import Session  # pragma: no cover
+
+
+class TempTableCleaner:
+    """
+    Clean up temporary tables created automatically by Snowpark at the session level.
+
+    This function identifies and removes any temporary tables that have been
+    generated by Snowpark during the current session. Temporary tables are
+    typically used for intermediate computations (e.g., df.cache_result) and
+    are not needed when they are not referenced anymore. Removing these tables
+    helps maintain a clean working environment and reduce storage cost for a long-running
+    session.
+    """
+
+    def __init__(self, session: "Session") -> None:
+        self.session = session
+        # this dict maintains key-value pair from Snowpark-generated temp table fully-qualified name
+        # to its reference count for later temp table management
+        # this dict will still be maintained even if the cleaner is stopped (`stop()` is called)
+        self.ref_count_map: Dict[str, int] = defaultdict(int)
+        self.queue = Queue()
+        self.cleanup_thread = None
+
+    def inc_ref_count(self, name: str) -> None:
+        self.ref_count_map[name] += 1
+
+    def dec_ref_count(self, name: str) -> None:
+        self.ref_count_map[name] -= 1
+        if self.ref_count_map[name] <= 0:
+            self.ref_count_map.pop(name)
```

Comment:
shall we pop the table now, or pop it out of the map after the drop is successfully done?

Reply:
It doesn't really matter for now, because `dec_ref_count` is only triggered during garbage collection once, so this table won't be visited again anyway. But if we want to consider [this case](https://github.com/snowflakedb/snowpark-python/pull/1983#discussion_r1693390798), we should pop it out only when the drop is successfully done.